src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -94,90 +79,196 @@ public class DistributedTransactionContext : ITransactionContext, IEnlistmentNot
-							if (sessionImplementor.FlushMode != FlushMode.Never && sessionImplementor.ConnectionManager.IsConnected)
+							_sessionImplementor.BeforeTransactionCompletion(null);
+							if (_sessionImplementor.FlushMode != FlushMode.Never && _sessionImplementor.ConnectionManager.IsConnected)
 							{
-								using (sessionImplementor.ConnectionManager.FlushingFromDtcTransaction)
+								using (_sessionImplementor.ConnectionManager.FlushingFromDtcTransaction)
 								{
-									logger.Debug(string.Format("[session-id={0}] Flushing from Dtc Transaction", sessionImplementor.SessionId));
-									sessionImplementor.Flush();
+									_logger.DebugFormat("[session-id={0}] Flushing from Dtc Transaction", _sessionImplementor.SessionId);
+									_sessionImplementor.Flush();
 								}
 							}
-							logger.Debug("prepared for DTC transaction");
+							_logger.Debug("prepared for DTC transaction");
 
 							tx.Complete();
 						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						_waitEvent.Reset();
+
 						preparingEnlistment.Prepared();
 					}
 					catch (Exception exception)
 					{
-						logger.Error("DTC transaction prepare phase failed", exception);
+						_logger.Error("DTC transaction prepare phase failed", exception);
 						preparingEnlistment.ForceRollback(exception);
 					}
 				}
 			}
 
 			void IEnlistmentNotification.Commit(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, true);
+
+			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, false);
+
+			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, null);
+
+			private void ProcessSecondPhase(Enlistment enlistment, bool? success)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
 				{
-					logger.Debug("committing DTC transaction");
-					// we have nothing to do here, since it is the actual
-					// DB connection that will commit the transaction
+					_logger.Debug(
+						success.HasValue
+							? success.Value
+								? "committing DTC transaction"
+								: "rolled back DTC transaction"
+							: "DTC transaction is in doubt");
+					// we have not much to do here, since it is the actual
+					// DB connection that will commit/rollback the transaction
+					// Usual cases will raise after transaction actions from TransactionCompleted event.
+					if (!success.HasValue)
+					{
+						// In-doubt. A durable ressource has failed and may recover, but we won't wait to know.
+						RunAfterTransactionActions(false);
+					}
+
 					enlistment.Done();
-					IsInActiveTransaction = false;
 				}
 			}
 
-			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+			#endregion
+
+			public void TransactionCompleted(object sender, TransactionEventArgs e)
```

Comment:
TransactionCompleted shall be incorporated to IEnlistmentNotification.Commit/IEnlistmentNotification.Rollback/IEnlistmentNotification.InDoubt